File: karate-gatling/src/test/scala/mock/CatsSimulation.scala
package mock

import com.intuit.karate.Runner
import com.intuit.karate.gatling.PreDef._
import io.gatling.core.Predef._

import scala.concurrent.duration._

class CatsSimulation extends Simulation {

  MockUtils.startServer(0)

  val protocol = karateProtocol(
    "/cats/{id}" -> Nil,
    "/cats" -> pauseFor("get" -> 15, "post" -> 25)
  )

  protocol.nameResolver = (req, ctx) => req.getHeader("karate-name")
  protocol.runner.karateEnv("perf")

  val create = scenario("create").exec(karateFeature("classpath:mock/cats-create.feature")).exec(session => {
    println("*** id in gatling: " + session("id").as[String])
    println("*** session status in gatling: " + session.status)
    session
  })
  val delete = scenario("delete").group("delete cats") {
    exec(karateFeature("classpath:mock/cats-delete.feature@name=delete"))
  }
  val custom = scenario("custom").exec(karateFeature("classpath:mock/custom-rpc.feature"))

  setUp(
    create.inject(rampUsers(10) during (5 seconds)).protocols(protocol),
    delete.inject(rampUsers(5) during (5 seconds)).protocols(protocol),
    custom.inject(rampUsers(10) during (5 seconds)).protocols(protocol)
  )

}

File: karate-gatling/src/test/java/mock/CatsSimulation.java
package mock;

import com.intuit.karate.gatling.javaapi.KarateProtocolBuilder;

import io.gatling.javaapi.core.ScenarioBuilder;
import io.gatling.javaapi.core.Simulation;

import static io.gatling.javaapi.core.CoreDsl.*;
import static com.intuit.karate.gatling.javaapi.KarateDsl.*;

public class CatsSimulation extends Simulation {

  public CatsSimulation() {
      
    MockUtils.startServer(0);

    KarateProtocolBuilder protocol = karateProtocol(
      uri("/cats/{id}").nil(),
      uri("/cats").pauseFor(method("get", 15), method("post", 25)
    ));

    protocol.nameResolver = (req, ctx) -> req.getHeader("karate-name");
    protocol.runner.karateEnv("perf");

    ScenarioBuilder create = scenario("create").exec(karateFeature("classpath:mock/cats-create.feature")).exec(session -> {
      System.out.println("*** id in gatling: " + session.getString("id"));
      System.out.println("*** session status in gatling: " + session.asScala().status());
      return session;
    });

    ScenarioBuilder delete = scenario("delete").group("delete cats").on(
      exec(karateFeature("classpath:mock/cats-delete.feature@name=delete"))
    );

    ScenarioBuilder custom = scenario("custom").exec(karateFeature("classpath:mock/custom-rpc.feature"));

    setUp(
      create.injectOpen(rampUsers(10).during(5)).protocols(protocol),
      delete.injectOpen(rampUsers(5).during(5)).protocols(protocol),
      custom.injectOpen(rampUsers(10).during(5)).protocols(protocol)
    );
  }
}
